lib/Email/Stuffer/TestLinks.pm

```diff
@@ -22,55 +26,74 @@ Email::Stuffer>send_or_die()
 
 =head1 DESCRIPTION
 
-When this module is included in a test, it parses HTML links (<a href="xyz"...)
-in every email sent through Email::Stuffer->send_or_die(). Each URI must get a
-successful response code (200 range) and the returned pagetitle must not contain
-'error' or 'not found'.
+When this module is included in a test, it parses http links (<a href="xyz">...</a>)
+and image links (<img src="xyz">) in every email sent through Email::Stuffer->send_or_die().
+Each URI must be  get a successful response code (200 range).
+Page title must not contain 'error' or 'not found' for text/html content.
+Image links must return an image content type.
 
 =cut
 
 install_modifier 'Email::Stuffer', after => send_or_die => sub {
 
     my $self = shift;
-    my $ua = Mojo::UserAgent->new(max_redirects => 10, connect_timeout => 5);
 
     my %urls;
     $self->email->walk_parts(
         sub {
             my ($part) = @_;
             return unless ($part->content_type && $part->content_type =~ /text\/html/i);
             my $dom = Mojo::DOM->new($part->body);
-            my $links = $dom->find('a')->map(attr => 'href')->compact;
+            push @{$urls{http}}, $dom->find('a')->map(attr => 'href')->compact->grep(sub {$_ !~ /^mailto:/})->uniq->to_array->@*;
+            push @{$urls{image}}, $dom->find('img')->map(attr => 'src')->compact->uniq->to_array->@*;
+    });
 
-            # Exclude anchors, mailto
-            $urls{$_} = 1 for (grep { !/^mailto:/ } @$links);
-        });
+    my @data = map { my $type = $_; map { [ $type, $_ ] } $urls{$type}->@* } keys %urls;
 
-    for my $url (sort keys %urls) {
+    my $loop = IO::Async::Loop->new();
```

Comment:
Nitpicking, coz this is a singlton, but couldn't it be moved to the module scope to be initialized once?

Reply:
I tried it and got this interesting error when running a RPC test:
```
Unexpected event loop class: had IO::Async::Loop::Epoll and IO::Async::Loop::Epoll from magic constructor for IO::Async, expected a subclass of IO::Async::Loop::Mojo at lib/BOM/RPC/Transport/HTTP.pm line 26.
```

Reply:
What was the test? Basically whoever calls `IO::Async::Loop->new` first wins, that means this module is included before the other one.

The correct way of using an event loop inside a module is to make it child of `IO::Async::Notifier` and add it to the loop in the consumer side.

Otherwise we will end up having the same issue with who called that `->new` first :)

Reply:
It seems to happen with any test that includes this module, e.g. /home/git/regentmarkets/bom-rpc/t/BOM/RPC/10_authorize.t
Some background: this test used to use Mojo::UserAgent to check URLs sequentially. It became quite slow after including images, so I changed it to Net::Async::HTTP so can fetch many URLs concurrently. Open to better ways of doing this.

Reply:
Just make it a child of `IO::Async::Notifier`, you can use [Binary::API](https://github.com/regentmarkets/perl-Binary-API/blob/master/lib/Net/Async/BinaryWS.pm#L6) as an example